### PR TITLE
first take on peering in CI

### PR DESCRIPTION
**Pull request by hawkowl**
### Which issue this PR addresses:

https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/7553173/

### What this PR does / why we need it:

private visibility + peering means we don't have to fix up NSGs to get CI working

### Test plan for issue:
hopefully e2e just works(tm)

### CODE REVIEW

File: pkg/util/cluster/cluster.go [hunk 1]
```diff
@@ -258,6 +266,25 @@ func (c *Cluster) Create(ctx context.Context, vnetResourceGroup, clusterName str
...
+	v3 := byte(v & 0xFF)
+	v2 := byte((v >> 8) & 0xFF)
+	v1 := byte((v >> 16) & 0xFF)
+	v0 := byte((v >> 24) & 0xFF)
+
+	vnetPrefix = fmt.Sprintf("%s/23", net.IPv4(v0, v1, v2, v3))
+	masterSubnet = fmt.Sprintf("%s/24", net.IPv4(v0, v1, v2, v3))
+	workerSubnet = fmt.Sprintf("%s/24", net.IPv4(v0, v1, v2+1, v3))
+	return
+}
```
Comment (jim-minter) [inline pkg/util/cluster/cluster.go:286]: is something like this less terrifying?
```golang
// pick a random /23 in the range [10.0.2.0, 10.128.0.0).  10.0.0.0 is used
// by dev-vnet to host CI; 10.128.0.0+ is used for pods.

var x, y int
for x == 0 && y == 0 {
	x, y = rand.Intn(128), 2*rand.Intn(128)
}

vnetPrefix = fmt.Sprintf("10.%d.%d.0/23", x, y)
masterSubnet = fmt.Sprintf("10.%d.%d.0/24", x, y)
workerSubnet = fmt.Sprintf("10.%d.%d.0/24", x, y+1)
```

File: pkg/util/cluster/cluster.go [hunk 2]
```diff
@@ -403,38 +430,33 @@ func (c *Cluster) createCluster(ctx context.Context, vnetResourceGroup, clusterN
 }
 
 func (c *Cluster) fixupNSGs(ctx context.Context, vnetResourceGroup, clusterName string) error {
```
Comment (jim-minter) [inline pkg/util/cluster/cluster.go:432]: fixupNSGs should be left as is and new function created for the peering.

File: pkg/util/cluster/cluster.go [hunk 3]
```diff
@@ -403,38 +430,33 @@ func (c *Cluster) createCluster(ctx context.Context, vnetResourceGroup, clusterN
...
-	var nsgs []mgmtnetwork.SecurityGroup
-	err := wait.PollImmediateUntil(10*time.Second, func() (bool, error) {
-		var err error
-		nsgs, err = c.securitygroups.List(ctx, "aro-"+clusterName)
-		return len(nsgs) > 0, err
-	}, timeoutCtx.Done())
-	if err != nil {
-		return err
+	clusterProp := &mgmtnetwork.VirtualNetworkPeeringPropertiesFormat{
+		RemoteVirtualNetwork: &mgmtnetwork.SubResource{
```
Comment (jim-minter) [inline pkg/util/cluster/cluster.go:437]: I expect you will probably want this here and below too:
```
AllowVirtualNetworkAccess: to.BoolPtr(true),
AllowForwardedTraffic:     to.BoolPtr(true),
```

(see function halfPeering)

File: pkg/util/cluster/cluster.go (hunk 2 above)
Comment (hawkowl) [reply pkg/util/cluster/cluster.go]: done

File: pkg/util/cluster/cluster.go (hunk 3 above)
Comment (hawkowl) [reply pkg/util/cluster/cluster.go:478]: done

File: pkg/util/cluster/cluster.go [hunk 4]
```diff
@@ -252,12 +261,36 @@ func (c *Cluster) Create(ctx context.Context, vnetResourceGroup, clusterName str
 		if err != nil {
 			return err
 		}
+
+		c.log.Info("peering subnets to CI infra")
+		err = c.peerSubnetsToCI(ctx, vnetResourceGroup, clusterName)
```
Comment (jim-minter) [inline pkg/util/cluster/cluster.go:266]: Looking at https://ms.portal.azure.com/#@microsoft.onmicrosoft.com/resource/subscriptions/46626fc5-476d-41ad-8c76-2ec49c6994eb/resourceGroups/e2einfra-eastus/providers/Microsoft.Network/virtualNetworks/dev-vnet/peerings I think there's going to need to be an (albeit best effort) deletion cleanup step

File: pkg/util/azureclient/mgmt/network/virtualnetworks.go [hunk 5]
```diff
@@ -33,3 +33,25 @@ func NewVirtualNetworksClient(environment *azure.Environment, subscriptionID str
 		VirtualNetworksClient: client,
 	}
 }
+
+type VirtualNetworkPeeringsClient interface {
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply pkg/util/azureclient/mgmt/network/virtualnetworks.go:37]: done

File: pkg/util/cluster/cluster.go (hunk 1 above)
Comment (hawkowl) [reply pkg/util/cluster/cluster.go]: Yeah, I went with this in the first place -- I guess it is much nicer, I was just hoping that there was a nicer option than string formatting, lol

File: pkg/util/cluster/cluster.go (hunk 4 above)
Comment (hawkowl) [reply pkg/util/cluster/cluster.go:266]: done

File: pkg/util/cluster/cluster.go [hunk 6]
```diff
@@ -289,6 +317,12 @@ func (c *Cluster) Delete(ctx context.Context, vnetResourceGroup, clusterName str
 				errs = append(errs, err)
 			}
 		}
+
+		// Do a best-effort attempt at deleting the network peering.
+		err = c.peerings.DeleteAndWait(ctx, c.env.ResourceGroup(), "dev-vnet", vnetResourceGroup+"-peer")
+		if err != nil {
+			c.log.Info("Couldn't delete network peering:", err)
```
Comment (jim-minter) [inline pkg/util/cluster/cluster.go:324]: sorry, should be `errs = append(errs, err)`

File: pkg/util/cluster/cluster.go [hunk 7]
```diff
@@ -437,6 +471,40 @@ func (c *Cluster) fixupNSGs(ctx context.Context, vnetResourceGroup, clusterName
...
+	}
+	rpProp := &mgmtnetwork.VirtualNetworkPeeringPropertiesFormat{
+		RemoteVirtualNetwork: &mgmtnetwork.SubResource{
+			ID: &cluster,
+		},
+		AllowVirtualNetworkAccess: to.BoolPtr(true),
+		AllowForwardedTraffic:     to.BoolPtr(true),
+	}
+
+	c.log.Infof("making peering from %s to %s", rp, cluster)
```
Comment (jim-minter) [inline pkg/util/cluster/cluster.go:493]: duplicate logline?

**Review by jim-minter — APPROVED** [review]

File: pkg/util/cluster/cluster.go [hunk 8]
```diff
@@ -357,12 +391,12 @@ func (c *Cluster) createCluster(ctx context.Context, vnetResourceGroup, clusterN
 				},
 			},
 			APIServerProfile: api.APIServerProfile{
-				Visibility: api.VisibilityPublic,
+				Visibility: api.VisibilityPrivate,
```
Comment (jim-minter) [inline pkg/util/cluster/cluster.go:394]: let's make this change dependent on CI for now and poll the team for what to do about it

File: pkg/util/azureclient/mgmt/network/virtualnetworkpeerings.go [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+	"context"
+
+	mgmtnetwork "github.com/Azure/azure-sdk-for-go/services/network/mgmt/2019-07-01/network"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/azure"
+)
+
+type VirtualNetworkPeeringsClient interface {
+	CreateOrUpdate(ctx context.Context, resourceGroupName string, virtualNetworkName string, virtualNetworkPeeringName string, virtualNetworkPeeringParameters mgmtnetwork.VirtualNetworkPeering) (result mgmtnetwork.VirtualNetworkPeeringsCreateOrUpdateFuture, err error)
+	Delete(ctx context.Context, resourceGroupName string, virtualNetworkName string, virtualNetworkPeeringName string) (result mgmtnetwork.VirtualNetworkPeeringsDeleteFuture, err error)
```
[the remark this answers is not in the document]
Comment (hawkowl) [reply pkg/util/azureclient/mgmt/network/virtualnetworkpeerings.go:16]: it's used by the addons client, isn't it?

**Review by mjudeikis — APPROVED** [review]

**Comment by mjudeikis** [comment]
Feel free to merge once you ready :)

**Pull request merged**